src/map_votes.cpp

```diff
@@ -0,0 +1,543 @@
+	int iMapIndexToVote = g_pGameRules->m_nEndMatchMapGroupVoteOptions[iVoteOption];
+	if (iMapIndexToVote < 0 || iMapIndexToVote >= m_vecMapList.Count()) return;
+
+	// Set the vote for the player
+	int iSlot = pController->GetPlayerSlot();
+	m_arrPlayerVotes[iSlot] = iVoteOption;
+
+	// Log vote to console
+	const char* sMapName = g_pMapVoteSystem->GetMapName(iMapIndexToVote);
+	Msg("Adding vote to map %i (%s) for player %s (slot %i).\n", iVoteOption, sMapName, pController->GetPlayerName(), iSlot);
+
+	// Update the winning map for every player vote
+	UpdateWinningMap();
+}
+
+bool CMapVoteSystem::UpdateWinningMap()
+{
+	int iWinningMapIndex = WinningMapIndex();
+	if (iWinningMapIndex >= 0) {
+		g_pGameRules->m_nEndMatchMapVoteWinner = iWinningMapIndex;
+		return true;
+	}
+	return false;
+}
+
+int CMapVoteSystem::WinningMapIndex()
+{
+	// Count the votes of every player
+	int arrMapVotes[10] = {0};
+	for (int i = 0; i < gpGlobals->maxClients; i++) {
+		auto pController = CCSPlayerController::FromSlot(i);
+		if (pController && pController->IsConnected() && m_arrPlayerVotes[i] >= 0) {
+			arrMapVotes[m_arrPlayerVotes[i]]++;
+		}
+	}
+
+	// Identify the max. number of votes
+	int iMaxVotes = 0;
+	for (int i = 0; i < 10; i++) {
+		iMaxVotes = (arrMapVotes[i] > iMaxVotes) ? arrMapVotes[i] : iMaxVotes;
+	}
+
+	// Identify how many maps are tied with the max number of votes    
+	int iMapsWithMaxVotes = 0;
+	for (int i = 0; i < 10; i++) {
+		if (arrMapVotes[i] == iMaxVotes) iMapsWithMaxVotes++;
+	}
+
+	// Break ties: 'random' map with the most votes
+	int iWinningMapTieBreak = m_iRandomWinnerShift % iMapsWithMaxVotes;
+	int iWinningMapCount = 0;
+	for (int i = 0; i < 10; i++) {
+		if (arrMapVotes[i] == iMaxVotes) {
+			if (iWinningMapCount == iWinningMapTieBreak) return i;
+			iWinningMapCount++;
+		}
+	}
+	return -1;
+}
+
+void CMapVoteSystem::GetNominatedMapsForVote(CUtlVector<int>& vecChosenNominatedMaps)
+{
+	int iNumDistinctMaps = 0;
+	CUtlVector<int> vecAvailableNominatedMaps;
+	for (int i = 0; i < gpGlobals->maxClients; i++) {
+		int iNominatedMapIndex = m_arrPlayerNominations[i];
+
+		// Introduce nominated map indexes and count the total number
+		if (iNominatedMapIndex != -1) {
+			if (!vecAvailableNominatedMaps.HasElement(iNominatedMapIndex)) {
+				iNumDistinctMaps++;
+			}
+			vecAvailableNominatedMaps.AddToTail(iNominatedMapIndex);
+		}
+	}
+
+	// Randomly select maps out of the set of nominated maps
+	// weighting by number of nominations, and returning a random order
+	int iMapsToIncludeInNominate = (iNumDistinctMaps < m_iMaxNominatedMaps) ? iNumDistinctMaps : m_iMaxNominatedMaps;
+	while (vecChosenNominatedMaps.Count() < iMapsToIncludeInNominate) {
+		int iMapToAdd = vecAvailableNominatedMaps[rand() % vecAvailableNominatedMaps.Count()];
+		vecChosenNominatedMaps.AddToTail(iMapToAdd);
+		while (vecAvailableNominatedMaps.HasElement(iMapToAdd)) {
+			vecAvailableNominatedMaps.FindAndRemove(iMapToAdd);
+		}
+	}
```

Comment:
You've commented this as intentional behaviour, so I have to ask, why are we randomly picking from the nominated maps? From what I can tell, the mini nomination voting system used these days (GFL mapchooser, mapchooser unlimited) was otherwise brought over other than this implementation difference.

Reply:
I wanted to allow weighted nominations in the 'base' implementation so that multiple players can nominate the same map and we randomly pick (with weights equal to the number of maps). Agree w/ implementing a 'select top K maps by number of noms and break ties' and 'do not allow multiple noms per map', but I'd do it as a second PR — the idea was that as far as I can tell, this is the most generic 'common' base.